frontend/src/utils/executor.ts

```diff
@@ -18,19 +56,55 @@ const getInstructions = async (code: string): Promise<instructionRes> => {
       process.env.REACT_APP_EXECUTOR_ENDPOINT + '/execute',
       {
         code,
+        input: '',
       },
     )
+    const ret: instructionRes = {}
     const instructions = res.data?.data
-    if (instructions) return { instructions }
-    return {
-      errorMessage: res.data?.errorMessage || defaultErrorMessage,
+    if (instructions) ret.instructions = instructions
+    if (!instructions || res.data?.error) {
+      ret.errorMessage = res.data?.error || defaultErrorMessage
     }
+    return ret
   } catch (err) {
     return {
       errorMessage: defaultErrorMessage,
     }
   }
 }
 
-export { getInstructions }
-export type { instruction }
+// This cannot be defined simply as one type, because Typescript has issues
+// with circular references.
+type parsedVariablePrimitive = boolean | number | string
+type parsedVariableArray = Array<parsedVariable>
+type parsedVariableRecord = { [key: string]: parsedVariable }
+type parsedVariable =
+  | parsedVariablePrimitive
+  | parsedVariableArray
+  | parsedVariableRecord
+
+// parses the data returned from executor into a string or number
+const parseVariableValue = (value: variableValue): parsedVariable => {
+  if (typeof value.value === 'string') {
+    return value.value
+  }
+
+  if (value.type === 'dict') {
+    const dict: parsedVariableRecord = {}
+    for (const key in value.value) {
+      dict[key] = parseVariableValue(value.value[key])
+    }
+    return dict
+  }
+
+  if (value.type === 'list' || value.type === 'set' || value.type === 'tuple') {
```

Comment:
As mentioned, would be better to use Enum instead of string literals. But is fine for now, to kiv